### ISSUE TITLE
'pythoncom' is not defined

### ISSUE THREAD

**Issue by solanu** [Nuitka/Nuitka#296]
I get the following error when I run `main.exe` after I compile with `python -m nuitka --standalone --mingw64 --follow-imports --enable-plugin=numpy-plugin main.py`:
```
Traceback (most recent call last):
  File "C:\Users\IEUser\Desktop\o\main.dist\main.py", line 24, in <module>
  File "C:\Users\IEUser\Desktop\o\main.dist\c\i.py", line 24, in __init__
  File "C:\Users\IEUser\Desktop\o\main.dist\c\opc.py", line 11, in __init__
  File "C:\Users\IEUser\Desktop\o\main.dist\OpenOPC.py", line 155, in __init__
NameError: name 'pythoncom' is not defined
```

My setup is:
```
Python: 3.7.2 (tags/v3.7.2:9a3ffc0492, Dec 23 2018, 23:09:28) [MSC v.1916 64 bit (AMD64)] (It is installed with Windows installer) [no more versions of python installed]
Nuitka 0.6.2 (it is installed with pip)
MinGW-64 8.1.0 (win32 threads and x86_64 architecture) (It is installed with Windows installer)
Windows 10 64bits [fresh and clean original installation]
Arch: x86_64
```
If I am right `pythoncom` is in `pywin32` module and I have installed it using `pip`.

**Comment by JorjMcKie** [comment]
As a Windows user, let me try to help.
Can you please give your script (or a skeleton of it) to reproduce the error: which pywin comonents did you you use, etc.

**Comment by solanu** [comment]
My (simplified) script `opc.py`:

```
import OpenOPC

class ClientOPC():
    def __init__(self,server,ip,user,pass,domain,clsid):
        self.server=server
        self.ip=ip
        self.user=user
        self.password=pass
        self.domain=domain
        self.clsid=clsid
        self.opc=OpenOPC.client() # the problem is here <==============
```

I am using OpenOPC (`pip install OpenOPC-Python3x`) and Pywin32 (`pip install pywin32`).

**Comment by JorjMcKie** [comment]
Have no internet since many hours.
Just putting this in my e-mail out basket, so it will get transmitted on first possibility.
Your code below is not sufficient to re-create the situation. It does get compiled without problems.
Please provide more setup info.

Jorj

________________________________
Von: solanu <notifications@github.com>
Gesendet: Wednesday, March 13, 2019 2:57:47 AM
An: Nuitka/Nuitka
Cc: Jorj X. McKie; Assign
Betreff: Re: [Nuitka/Nuitka] 'pythoncom' is not defined (#296)


My (simplified) script opc.py:

import OpenOPC

class ClientOPC():
    def __init__(self,server,ip,user,pass,domain,clsid):
        self.server=server
        self.ip=ip
        self.user=user
        self.password=pass
        self.domain=domain
        self.clsid=clsid
        self.opc=OpenOPC.client() # the problem is here <==============


I am using OpenOPC (pip install OpenOPC-Python3x) and Pywin32 (pip install pywin32).

**Comment by solanu** [comment]
@JorjMcKie I have provided all setup info in my first post and I have added more information after that.

Anyway, if I copy pythoncom37.dll into main.dist directory, it works. Is there any any way to compile and copy that dll automatically?

**Comment by JorjMcKie** [comment]
Hi, I am back online.
I have another script that uses pythoncom for some other purpose. It compiles & executes flawlessly with standalone.
Can you please try compiling [link-maker.py](https://github.com/Nuitka/NUITKA-Utilities/blob/master/link-maker.py)? It is a little utility which finds ``.exe`` files in a folder and puts ``.lnk`` links for them on your ``Desktop``.

If you compile it with these options:

```python -m nuitka --standalone --mingw64 --python-flag=nosite --experimental=use_pefile --disable-dll-dependency-cache ...```

without getting a working executable, then something must be wrong with your pywin / pythoncom installation. Maybe your should run again the post-installation step of it.

So my hope is that using ``use_pefile`` does a better job in determining that this dll is indeed needed.

Apart from all this, there is always the possibility to write a little user plugin that does the copy ... but it really should not be necessary.

**Comment by JorjMcKie** [comment]
If you find the above invocation string a little lengthy: try [nuitka-sa.py](https://github.com/Nuitka/NUITKA-Utilities/blob/master/nuitka-sa.py). It does all of the above internally and you can invoke with ``python nuitka-sa.py yourscript.py``.

**Comment by solanu** [comment]
> Can you please try compiling [link-maker.py](https://github.com/Nuitka/NUITKA-Utilities/blob/master/link-maker.py)? It is a little utility which finds `.exe` files in a folder and puts `.lnk` links for them on your `Desktop`.
> 
> If you compile it with these options:
> 
> `python -m nuitka --standalone --mingw64 --python-flag=nosite --experimental=use_pefile --disable-dll-dependency-cache ...`
> 
> without getting a working executable, then something must be wrong with your pywin / pythoncom installation. Maybe your should run again the post-installation step of it.
> 
> So my hope is that using `use_pefile` does a better job in determining that this dll is indeed needed.

Compilation works fine but when I run `link-maker.exe` file I get the following: 

```
Traceback (most recent call last):
  File "C:\Users\IEUser\Desktop\test\link-maker.dist\link-maker.py", line 22, in <module>
ModuleNotFoundError: No module named 'PySimpleGUI'
```

**Comment by JorjMcKie** [comment]
Please also install ``pip install pysimplegui``.
It is a tkinter based package for dead easy GUI creation, which link-maker uses to ask for the desired folder.

**Comment by JorjMcKie** [comment]
In parallel you could also try compiling your problem script with the parameters I named a few posts upwards.

**Comment by JorjMcKie** [comment]
I don't know how far we will get with these tries.
If they fail, then we may have to do an additional "implicit import" that simulates that

> Magic utility that "redirects" to pythoncomxx.dll
import pywintypes
pywintypes.__import_pywin32_system_module__("pythoncom", globals())

Whatever is required ... we will make it!

In order to get your situation running on my computer, can you give me your `main.py`?

**Comment by solanu** [comment]
> Please also install `pip install pysimplegui`.


```
Traceback (most recent call last):
  File "C:\Users\IEUser\Desktop\test\link-maker.dist\link-maker.py", line 40, in <module>
  File "C:\Users\IEUser\Desktop\test\link-maker.dist\PySimpleGUI\PySimpleGUI.py", line 3720, in Read
  File "C:\Users\IEUser\Desktop\test\link-maker.dist\PySimpleGUI\PySimpleGUI.py", line 3645, in Show
  File "C:\Users\IEUser\Desktop\test\link-maker.dist\PySimpleGUI\PySimpleGUI.py", line 6017, in StartupTK
  File "C:\Users\IEUser\Desktop\test\link-maker.dist\tkinter\__init__.py", line 2023, in __init__
_tkinter.TclError: Can't find a usable init.tcl in the following directories:
    C:/Users/IEUser/Desktop/prueba/lib/tcl8.6 C:/Users/IEUser/Desktop/test/lib/tcl8.6 C:/Users/IEUser/Desktop/lib/tcl8.6 C:/Users/IEUser/Desktop/test/library C:/Users/IEUser/Desktop/library C:/Users/IEUser/Desktop/tcl8.6.8/library C:/Users/IEUser/tcl8.6.8/library

This probably means that Tcl wasn't installed properly.
```

**Comment by JorjMcKie** [comment]
:-) sorry: you must now also use option ``--enable-plugin=tk-plugin`` in your compile

**Comment by solanu** [comment]
> In parallel you could also try compiling your problem script with the parameters I named a few posts upwards.

It does not work, it does not copy pythoncom37.dll into `main.dist` directory but don't worry. I can copy it manually.

**Comment by solanu** [comment]
> :-) sorry: you must now also use option `--enable-plugin=tk-plugin` in your compile

It works with `link-maker` program but it does not work with `pythoncom`. Don't worry, it is ok.

**Comment by JorjMcKie** [comment]
@solanu ... you are welcome.
However, I hate open ended things. Here is my suggestion:
Obviously, the `OpenOPC` package is tricking out the normal DLL detection (for which reason they themselves have this _**magic**_ ...).
Here is a NUITKA user plugin that should achieve the same thing:
[add-pythoncom.zip](https://github.com/Nuitka/Nuitka/files/2967464/add-pythoncom.zip)

Your compilation command should include it like this:

```
python -m nuitka --standalone ... --user-plugin=add-pythoncom.py ... yourscript.py
```
The plugin detects that your script needs ``pythoncomXX.dll`` and whether the final dist folder does not contain it. In that case it is copied over (as you would have done manually).

Should be a satisfactory solution for the time being. In the end I imagine we extend Nuitka's standard plugins to take over this task.

Please let me know what you think.

**Comment by solanu** [comment]
> Here is a NUITKA user plugin that should achieve the same thing:
> [add-pythoncom.zip](https://github.com/Nuitka/Nuitka/files/2967464/add-pythoncom.zip)
> 
> Your compilation command should include it like this:
> 
> ```
> python -m nuitka --standalone ... --user-plugin=add-pythoncom.py ... yourscript.py
> ```

It does not works. `add-pythoncom.py` plugin does not copy `pythoncom37.dll` to `main.dist`.

**Comment by solanu** [comment]
> But `pywintypes37.dll` is in `main.dist`? 

Yes, it is but Nuitka always copies `pywintypes37.dll` to `main.dist`, it is not necessary to use `--user-plugin` option for that. The problem is with `pythoncom37.dll`.

> And `pywintypes37.dll` and `pythoncom37.dll` both in the same folder `...\Python37\Lib\site-packages\pywin32_system32`?

Yes, they are.

**Comment by JorjMcKie** [comment]
Then I do not understand why this doesn't work. It looks exactly like my configuration.

The plugin **_uses_** the fact ``pywintypes.dll`` is detected and copied:
* Nuitka passes all DLL names to the plugin, that it wants to output to the dist folder
* when pywintypes.dll is presented, the plugin notes that fact, and
* at the end also copies pythoncom to dist if that has not happened before.

When pywintypes is encountered, it saves its folder name part and appends "37.dll" to the string "pythoncom" (thus making sure the correct Python version is used).

Would you be willing to help find out what is happening and insert the following helper instructions in the plugin:

```python
import os
import sys
import shutil
from logging import info

from nuitka import Options
from nuitka.plugins.PluginBase import UserPluginBase


class Usr_Plugin(UserPluginBase):

    plugin_name = __file__

    def __init__(self):
        self.pywin32_dir = None
        self.pywin32_com = "pythoncom"
        self.do_copy = True
        return None

    def removeDllDependencies(self, dll_filename, dll_filenames):
        """ CHeck if we are compiling a pythoncom script.
        """
        dll_dir = os.path.dirname(dll_filename).lower()
        dll_file = os.path.basename(dll_filename).lower()

        if (
            dll_file.startswith("pywintypes")
            and dll_file.endswith(".dll")
            and self.pywin32_dir is None
        ):  # this script uses pythoncom
            self.pywin32_dir = dll_dir
            self.pywin32_com += dll_file[10:]
      ===>  info(" detected: " + self.pywin32_com) # this message appears on detecting pywintypes

        # the following might be deleted alltogether ...
        elif dll_file.startswith(self.pywin32_com) and dll_file.endswith(".dll"):
            # skip extra copy if pythoncomxx.dll already detected
            self.do_copy = False
       ===> info(" suppressing copy of " + self.pywin32_com) # we suppress extra copy if pythoncom

        return ()

    def onStandaloneDistributionFinished(self, dist_dir):
        """ Copy pythoncom DLL if needed."""

        if not self.do_copy:
            return None

        dll_from = os.path.join(self.pywin32_dir, self.pywin32_com)
        dll_to = os.path.join(dist_dir, self.pywin32_com)
        info(" Copying " + self.pywin32_com)
        shutil.copy(dll_from, dll_to, follow_symlinks=False)
        return None
```
At a minimum, the modified plugin should announce the " detected " message

**Comment by solanu** [comment]
Here is the ouput:
```
Invoking NUITKA with these options:
 --standalone
 --mingw64
 --python-flag=nosite
 --remove-output
 --experimental=use_pefile
 --disable-dll-dependency-cache
 --assume-yes-for-download
 --follow-imports
 --enable-plugin=numpy-plugin
 --enable-plugin=tk-plugin
 --user-plugin=add-pythoncom.py
 main.py

Nuitka:INFO:User plugin 'add-pythoncom.py' loaded.
Nuitka:INFO: Now copying tkinter libraries.
Nuitka:INFO: Finished copying tkinter libraries.
Nuitka:INFO: Now copying extra binaries from 'numpy' installation:
Nuitka:INFO: c:\program files\python37\lib\site-packages\numpy\.libs\libopenblas.csrrd7hkrkc3t3yxa7vy7tazglswdkw6.gfortran-win_amd64.dll
Nuitka:INFO: Copied 1 binary.
Nuitka:INFO: detected: pythoncom37.dll
```

I will be out of the office, I am coming back on Wednesday

**Comment by JorjMcKie** [comment]
I probably was too picky. Here is another version, which copies everything that is in ``pywintypesXX.dll``. Obviously just 2 small files.

[add-pythoncom.zip](https://github.com/Nuitka/Nuitka/files/2971126/add-pythoncom.zip)

**Comment by JorjMcKie** [comment]
I am taking @deajan on this thread:
Do you think it makes sense to include this in your pefile logic? You already have a list of DLLs you are checking against.
Another - and maybe better - option would be to include it in the standard plugins, probably ``ImplicitImports.py``.

**Comment by JorjMcKie** [comment]
@solanu - in the meantime, I have looked deeper into the code with respect to DLL dependency detection.
You can get along without the aformentioned user plugin with an additional option ``--experimental=use_pefile_recurse``. This also copies all DLLs in folder ``...\Python37\Lib\site-packages\pywin32_system32``. However at the cost of a somewhat higher execution time overall.
The synopsis of parameters leading to success in this case is:

```
python nuitka-sa2.py opc-test.py

NUITKA is compiling 'opc-test.py' with these options:
 --standalone
 --mingw64
 --python-flag=nosite
 --remove-output
 --experimental=use_pefile
 --experimental=use_pefile_recurse
 --disable-dll-dependency-cache
```
As mentioned before:
Requiring ``pythoncom37.dll`` by principle is **not detectable** via DLL dependency walkers. So, triggering some extra logic is needed one way or the other. The fastest way certainly is special casing via a user plugin or inclusion in one of the Nuitka standard plugins.
Timing difference between using the user plugin
```
NUITKA is compiling 'opc-test.py' with these options:
 --standalone
 --mingw64
 --python-flag=nosite
 --remove-output
 --experimental=use_pefile
 --disable-dll-dependency-cache
 --user-plugin=add-pythoncom.py
```
 and the above invocation parameters was 146 vs. 228 seconds for a mini-script on my machine.

**Comment by solanu** [comment]
> 
> 
> I probably was too picky. Here is another version, which copies everything that is in the folder of `pywintypesXX.dll`. Obviously just 2 small files.
> 
> [add-pythoncom.zip](https://github.com/Nuitka/Nuitka/files/2971126/add-pythoncom.zip)

Output (removing no interesting warnings):

```
Invoking NUITKA with these options:
 --standalone
 --mingw64
 --python-flag=nosite
 --remove-output
 --experimental=use_pefile
 --disable-dll-dependency-cache
 --assume-yes-for-download
 --follow-imports
 --enable-plugin=numpy-plugin
 --enable-plugin=tk-plugin
 --user-plugin=add-pythoncom.py
 main.py

Nuitka:INFO:User plugin 'add-pythoncom.py' loaded.
Nuitka:INFO: Now copying tkinter libraries.
Nuitka:INFO: Finished copying tkinter libraries.
Nuitka:INFO: Now copying extra binaries from 'numpy' installation:
Nuitka:INFO: c:\program files\python37\lib\site-packages\numpy\.libs\libopenblas.csrrd7hkrkc3t3yxa7vy7tazglswdkw6.gfortran-win_amd64.dll
Nuitka:INFO: Copied 1 binary.
Nuitka:INFO: detected pywintypes37.dll
```

It does not copy `pythoncom37.dll` to `main.dist` folder.

**Comment by solanu** [comment]
> python nuitka-sa2.py opc-test.py
> 
> NUITKA is compiling 'opc-test.py' with these options:
>  --standalone
>  --mingw64
>  --python-flag=nosite
>  --remove-output
>  --experimental=use_pefile
>  --experimental=use_pefile_recurse
>  --disable-dll-dependency-cache

It does not compile. Maybe because we have removed `--follow-imports` option or perhaps for having added `--experimental=use_pefile_recurse` option.

```
Invoking NUITKA with these options:
 --standalone
 --mingw64
 --python-flag=nosite
 --remove-output
 --experimental=use_pefile
 --experimental=use_pefile_recurse
 --disable-dll-dependency-cache
 --assume-yes-for-download
 --enable-plugin=numpy-plugin
 --enable-plugin=tk-plugin
 main.py

Nuitka:INFO: Now copying tkinter libraries.
Nuitka:INFO: Finished copying tkinter libraries.
Nuitka:INFO: Now copying extra binaries from 'numpy' installation:
Nuitka:INFO: c:\program files\python37\lib\site-packages\numpy\.libs\libopenblas.csrrd7hkrkc3t3yxa7vy7tazglswdkw6.gfortran-win_amd64.dll
Nuitka:INFO: Copied 1 binary.
Traceback (most recent call last):
  File "C:\Program Files\Python37\lib\site-packages\nuitka\__main__.py", line 184, in <module>
    main()
  File "C:\Program Files\Python37\lib\site-packages\nuitka\__main__.py", line 177, in main
    MainControl.main()
  File "C:\Program Files\Python37\lib\site-packages\nuitka\MainControl.py", line 787, in main
    standalone_entry_points=standalone_entry_points,
  File "C:\Program Files\Python37\lib\site-packages\nuitka\freezer\Standalone.py", line 1743, in copyUsedDLLs
    shutil.copyfile(dll_filename, target_path)
  File "C:\Program Files\Python37\lib\shutil.py", line 121, in copyfile
    with open(dst, 'wb') as fdst:
OSError: [Errno 22] Invalid argument: 'main.dist\\win32ui.pyd'
```

**Comment by JorjMcKie** [comment]
this is more than weird!

Anyway: the pywintypes DLL is detected correctly with the user plugin, but the DLLs are still not copied at end of dist folder creation.
This can only mean that you **have not installed** the latest Nuitka development version! It contains a critical method invoked at the end of standalone processing, which allows plugins to perform final housekeeping.

**Comment by solanu** [comment]
Yes, you are right, I have not installed the lastest development version of Nuitka.

**Comment by solanu** [comment]
Hi @JorjMcKie , compilations is successfully, dll is copied, but read from OPC is not working.

**Comment by JorjMcKie** [comment]
Currently I cannot run any OpenOPC script at all. I used to be successful with the following snippet:

```python
import OpenOPC

opc = OpenOPC.client()
opc.connect("Matrikon.OPC.Simulation")
print(opc["Square Waves.Real8"])
opc.close()
```

But now it says
```
Traceback (most recent call last):
  File "d:\Jorj\Desktop\Develop\nuitka\opc-test.py", line 3, in <module>
    opc = OpenOPC.client()
  File "C:\Users\Jorj\AppData\Local\Programs\Python\Python37\lib\site-packages\OpenOPC.py", line 155, in __init__
    pythoncom.CoInitialize()
NameError: name 'pythoncom' is not defined
```

Which imports are you using, resp. what does your script look like?

**Comment by kayhayen** [comment]
This has never gotten a good reproducer, but pythoncom has been solved in the mean time, and the crash was a bug deal t with in another issue.
